sycl/doc/design/PropertySets.md

```diff
@@ -0,0 +1,245 @@
+
+__Key:__ Specialization constant name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ Information about the specialization constant with the following
+fields: 
+
+```c++
+// Encodes ID of a scalar specialization constants which is a leaf of some
+// composite specialization constant.
+unsigned ID;
+// Encodes offset from the beginning of composite, where scalar resides, i.e.
+// location of the scalar value within a byte-array containing the whole
+// composite specialization constant. If descriptor is used to represent a
+// whole scalar specialization constant instead of an element of a composite,
+// this field should be contain zero.
+unsigned Offset;
+// Encodes size of scalar specialization constant.
+unsigned Size;
+```
+
+
+#### SYCL/specialization constants default values
+
+__Key:__ Specialization constant name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ Byte representation of the default value for the specialization
+constant.
+
+
+#### SYCL/devicelib req mask
+
+__Key:__ At most one entry with "DeviceLibReqMask".
+
+__Value type:__ 32 bit integer. ("1")
+
+__Value:__ A bitmask of which device libraries the binary uses.
+
+
+#### SYCL/kernel param opt
+
+__Key:__ Kernel name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ A bitmask identifying the arguments of the kernel that have been
+removed by the dead-argument-elimination optimization pass.
+
+
+#### SYCL/program metadata
+
+__Key:__ An arbitrary metadata key. This is often some identifier, such as a
+kernel name, followed by a '@' and some metadata identifier.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ Unspecified. Depends on the metadata key.
+
+
+#### SYCL/misc properties
+
+Miscellaneous properties:
+
+| Key                             | Value type            | Value                                                                                         |
+| ------------------------------- | --------------------- | --------------------------------------------------------------------------------------------- |
+| "isEsimdImage"                  | 32 bit integer. ("1") | 1 if the image is ESIMD and 0 otherwise.                                                      |
+| "sycl-register-alloc-mode"      | 32 bit integer. ("1") | The register allocation mode: 0 for automatic and 2 for large.                                |
+| "sycl-grf-size"                 | 32 bit integer. ("1") | The GRF size. Automatic if 0.                                                                 |
+| "optLevel"                      | 32 bit integer. ("1") | Optimization level, corresponding to the `-O` option used during compilation.                 |
+| "sanUsed"                       | Byte array. ("2")     | Specifying if address sanitization ("asan") or memory sanitization ("msan") is used.          |
+| "specConstsReplacedWithDefault" | 32 bit integer. ("1") | 1 if the specialization constants have been replaced by their default values and 0 otherwise. |
```

Comment:
Here and elsewhere ... are all of these keys required, or is it OK for some of them to be missing?  If a key is missing, what is the default value?

In general, it would be good to document whether keys are required vs. optional.

Reply:
To my knowledge, no properties are required. A lot of these have implementation (potentially L0) defined semantics if missing.

Reply:
I have made it a little more specific about what the behavior is when the various properties are missing.